RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc

```diff
@@ -0,0 +1,64 @@
+// Implementation of the mustache parameters interface
+
+#include "RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h"
+
+#include <algorithm>
+#include <utility>
+
+using namespace reco;
+
+MustacheSCParametersHelper::MustacheSCParametersHelper(const EcalMustacheSCParameters &params)
+    : EcalMustacheSCParameters(params) {}
+
+MustacheSCParametersHelper::MustacheSCParametersHelper(const edm::ParameterSet &iConfig) {
+  sqrtLogClustETuning_ = iConfig.getParameter<double>("sqrtLogClustETuning");  //1.1
+
+  // parabola parameters
+  const auto parabolaPSets = iConfig.getParameter<std::vector<edm::ParameterSet>>("parabolaParameterSets");
+  for (const auto &pSet : parabolaPSets) {
+    EcalMustacheSCParameters::ParabolaParameters params = {pSet.getParameter<double>("log10EMin"),
+                                                           pSet.getParameter<double>("etaMin"),
+                                                           pSet.getParameter<std::vector<double>>("pUp"),
+                                                           pSet.getParameter<std::vector<double>>("pLow"),
+                                                           pSet.getParameter<std::vector<double>>("w0Up"),
+                                                           pSet.getParameter<std::vector<double>>("w1Up"),
+                                                           pSet.getParameter<std::vector<double>>("w0Low"),
+                                                           pSet.getParameter<std::vector<double>>("w1Low")};
+    addParabolaParameters(params);
+    sortParabolaParametersCollection();
+  }
+}
+
+float MustacheSCParametersHelper::sqrtLogClustETuning() const { return sqrtLogClustETuning_; }
+
+void MustacheSCParametersHelper::setSqrtLogClustETuning(const float sqrtLogClustETuning) {
+  sqrtLogClustETuning_ = sqrtLogClustETuning;
+}
+
+EcalMustacheSCParameters::ParabolaParameters MustacheSCParametersHelper::parabolaParameters(float log10ClustE,
+                                                                                            float absSeedEta) const {
+  // assume the collection is sorted in descending ParabolaParameters.etaMin and descending ParabolaParameters.minEt
+  for (const auto &parabolaParams : parabolaParametersCollection_) {
+    if (log10ClustE < parabolaParams.log10EMin || absSeedEta < parabolaParams.etaMin) {
```

Comment:
How many elements `parabolaParametersCollection_` is expected to have? (just thinking if binary search would be worth it)

Reply:
I was wondering if there is a more efficient way of doing this. We have a set parameters in the pipeline (not in this PR) that would have 15*15 parameters. I think this will be fine grained enough for the future.

Reply:
Thanks. 225 elements sounds like a binary search with `std::lower_bound` should be faster on average.

Reply:
How would the search have to be set up in this case? `std::lower_bound` by default gives back an iterator to the element where `value <= element`. In particular if value > the largest element in the collection the iterator points to end(). What I would need is `element <= value` such that e.g. with two sets of parameters binned in eta (set 1 has etaMin = 0 and set 2 has etaMin = 1.5) for a value of eta = 2.1 I get back the second set and for eta = 0.3 I get back the first set.

I tried the following but that leads to the above mentioned problem.
```
  const auto vars = std::make_pair(log10ClustE, absSeedEta);                                                                                                  
  const auto it = std::lower_bound(parabolaParametersCollection_.begin(),                                                                                     
                                   parabolaParametersCollection_.end(),                                                                                       
                                   vars,                                                                                                                      
                                   [](const EcalMustacheSCParameters::ParabolaParameters &params,                                                             
                                      const std::pair<double, double> vars) {                                                                                 
                                     const auto bounds = std::make_pair(params.log10EMin, params.etaMin);                                                     
                                     return bounds > vars;                                                                                                    
                                   });                                                                                                                        
  return (it != parabolaParametersCollection_.end()) ? &*it : nullptr;
```

Reply:
After playing a bit with a toy code I think I see what you mean, i.e. `lower_bound()` returns an `element <= value`, but when the energy of `value` is larger than largest `energy` of `element`s, the eta of `element` can be larger than eta of  `value`. Even in that case the `element` found by `lower_bound()` could be used for a starting point of a linear search to find the correct box.

I'm not sure if the added complexity is worth of the effort though, so the search probably best to leave as it is unless a profiler shows it to be expensive.

Reply:
I was able to get it to work with a two step technique. The first search finds the right energy bin and the second search then looks for the matching eta bin. For the second search the range is already reduced since one does not need to look beyond the matched energy bins.